### PR TITLE
Fix and add warnings for RTL with invalid landing

### PR DESCRIPTION
**Pull request by Kjkinney**
**Describe problem solved by this pull request**
Currently RTL skips this segment of code which displays an error if the vehicle is unable to use the mission.:
`} else {
				// Otherwise use regular RTL.
				mavlink_log_critical(_navigator->get_mavlink_log_pub(), "RTL: unable to use mission landing");
			}`
We have also ran into issues with users missing this initial message. This has lead to users being confused as to why the fixedwing vehicle is not landing. 

**Describe your solution**
This gets rid of the dead code introduced previously, and fixes the bug with the warning message not being displayed. It also adds a message that displays every 30 seconds warning the user that the vehicle cannot use the mission landing. 

**Describe possible alternatives**
The best alternative would be to reject invalid missions and keep the current valid mission on the vehicle. This would get rid of an invalid landing being uploaded to the vehicle.  

**Test data / coverage**
This was tested in SITL
https://review.px4.io/plot_app?log=cfa987ce-62c4-4a50-b018-ca3456d892c3

A version of this has also been in use on our Fixedwing vehicles since August.

### CODE REVIEW

**Comment by Kjkinney** [comment]
FYI @Antiheavy

**Comment by Antiheavy** [comment]
FYI @sfuhrer I'm not 100% sure, but I think the bug with the warning message being bypassed got introduced in this PR: https://github.com/PX4/Firmware/pull/13320/files

We hope this PR by @Kjkinney remedies the bug and also improves the overall functionality of the warning message(s) for RTL_TYPE=1

**Comment by Antiheavy** [comment]
FYI @hamishwillee

**Comment by hamishwillee** [comment]
@Antiheavy Thanks for letting me know.

I'm not sure about this. The first check is  `_rtl_state > RTL_STATE_CLIMB`. You shouldn't be trying the landing while still climbing to the landing height. With this change won't the code immediately and always go to return to home landing if return is invoked below the return height? Could be wrong!!!

I'd also have thought we should check that the mission landing is defined when it gets selected as the destination - ie before this code. If it isn't valid we shouldn't get to this point.

Anyway, the docs for this currently say:

>Mission landing/rally point return (RTL_TYPE=1): Ascend to a safe altitude, fly direct to the closest destination other than home: rally point or start of mission landing. If no mission landing or rally points are defined, return home via direct path.

I think this is OK, though perhaps the last sentence should be "If no *valid* mission landing ..."

**Comment by Antiheavy** [comment]
FYI @julianoes  this was mentioned in the Dev call.

File: src/modules/navigator/rtl.cpp [hunk 1]
```diff
@@ -224,6 +229,15 @@ RTL::on_active()
 		advance_rtl();
 		set_rtl_item();
 	}
+
+	if (rtl_failed && !_navigator -> get_land_detected() -> landed) {
```
Comment (dagar) [inline src/modules/navigator/rtl.cpp]:
```suggestion
	if (rtl_failed && !_navigator->get_land_detected()->landed) {
```

File: src/modules/navigator/rtl.h [hunk 2]
```diff
@@ -131,6 +131,10 @@ class RTL : public MissionBlock, public ModuleParams
 	float _rtl_alt{0.0f};	// AMSL altitude at which the vehicle should return to the home position
 	bool _rtl_alt_min{false};
 
+	bool rtl_failed{false};
```
Comment (dagar) [inline src/modules/navigator/rtl.h]:
```suggestion
	bool _rtl_failed{false};
```

File: src/modules/navigator/rtl.h [hunk 3]
```diff
@@ -131,6 +131,10 @@ class RTL : public MissionBlock, public ModuleParams
...
+	bool warned{false};
```
Comment (dagar) [inline src/modules/navigator/rtl.h]:
```suggestion
	bool _warned{false};
```

File: src/modules/navigator/rtl.cpp [hunk 4]
```diff
@@ -257,10 +279,10 @@ RTL::set_rtl_item()
 
 	position_setpoint_triplet_s *pos_sp_triplet = _navigator->get_position_setpoint_triplet();
 
-	// Check if we are pretty close to the destination already.
+// Check if we are pretty close to the destination already.
```
Comment (dagar) [inline src/modules/navigator/rtl.cpp]:
```suggestion
	// Check if we are pretty close to the destination already.
```

**Comment by sfuhrer** [comment]
@Kjkinney I think you're correct that [this section](https://github.com/PX4/Firmware/blob/23e17aec42fedd4b66bd0b6ac584e7f0cdf7ef4f/src/modules/navigator/rtl.cpp#L247-L250) isn't actually used in any circumstance, because how RTL are handled currently, it automatically switches to uses destination.type = RTL_DESTINATION_HOME when the land start marker is missing. It will then loiter above home if a fixed-wing. 
If you want to have a more alerting message being sent to the user (like "RTL landing invalid" or something) I would add it [here](https://github.com/PX4/Firmware/blob/23e17aec42fedd4b66bd0b6ac584e7f0cdf7ef4f/src/modules/navigator/rtl.cpp#L183) when RTL_TYPE set by the user and the _destination.type do not match. And then also publish the "Upload a valid mission to use mission landing" messages periodically.

**Comment by julianoes** [comment]
@sfuhrer is this still relevant?

**Comment by sfuhrer** [comment]
> @sfuhrer is this still relevant?

Don't think so, as we do have that in place now afaik:

>The best alternative would be to reject invalid missions and keep the current valid mission on the vehicle. This would get rid of an invalid landing being uploaded to the vehicle.

**Comment by julianoes** [comment]
Ok, I'm going to close this then. Sorry about that @Kjkinney but thanks for the contribution!

**Pull request closed without merging**